projects/lib/src/board/extinctionboard.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+    This file is part of Cute Chess.
+
+    Cute Chess is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Cute Chess is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Cute Chess.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "extinctionboard.h"
+
+namespace Chess {
+
+ExtinctionBoard::ExtinctionBoard(const QString &set,
+				 bool  allPromotions)
+	: StandardBoard(),
+	m_allPromotions(allPromotions)
+{
+	for (auto s: set)
+		m_pieceSet.append(pieceFromSymbol(s));
+}
+
+Board* ExtinctionBoard::copy() const
+{
+	return new ExtinctionBoard(*this);
+}
+
+QString ExtinctionBoard::variant() const
+{
+	return "extinction";
+}
+
+QString ExtinctionBoard::defaultFenString() const
+{
+	return "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1";
+}
+
+bool ExtinctionBoard::kingsCountAssertion(int, int) const
+{
+	return extinctPiece(Side::White).isEmpty()
+	||     extinctPiece(Side::Black).isEmpty();
+}
+
+void ExtinctionBoard::addPromotions(int sourceSquare,
+				int targetSquare,
+				QVarLengthArray<Move>& moves) const
+{
+	if (m_allPromotions)
+		StandardBoard::addPromotions(sourceSquare, targetSquare, moves);
+	moves.append(Move(sourceSquare, targetSquare, King));
+}
+
+bool ExtinctionBoard::inCheck(Side, int) const
+{
+	return false;
+}
+
+Piece ExtinctionBoard::extinctPiece(Side side) const
+{
+	for (Piece pc: m_pieceSet)
+	{
+		int cnt = 0;
+		for (int i = 0; cnt < 1 && i < arraySize(); i++)
```

Comment:
This is clever, but it mislead me a bit because I thought we were counting things when we actually just wanted to find out if there's at least one piece on the board. I think it would be cleaner with a boolean variable and also faster because we could avoid the `cnt < 1` comparison at every step:

```
bool pieceFound = false;
for (int i = 0; i < arraySize(); i++)
{
	Piece piece = pieceAt(i);
	if (side == piece.side() && pc.type() == piece.type())
	{
		pieceFound = true;
		break;
	}
}
if (!pieceFound)
	return Piece(side, pc.type());
```